File: attocode/src/providers/adapters/openai.ts
/**
 * OpenAI Provider Adapter
 *
 * Adapts the OpenAI API to our LLMProvider interface.
 * Supports native tool use via the chatWithTools method.
 */

import type {
  LLMProvider,
  LLMProviderWithTools,
  Message,
  MessageWithContent,
  ChatOptions,
  ChatOptionsWithTools,
  ChatResponse,
  ChatResponseWithTools,
  ToolCallResponse,
  ToolDefinitionSchema,
  OpenAIConfig,
} from '../types.js';
import { ProviderError } from '../types.js';
import { registerProvider, hasEnv, requireEnv } from '../provider.js';
import { resilientFetch, type NetworkConfig } from '../resilient-fetch.js';

// =============================================================================
// OPENAI API TYPES
// =============================================================================

/** OpenAI message format */
interface OpenAIMessage {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content: string | null;
  name?: string;
  tool_calls?: OpenAIToolCall[];
  tool_call_id?: string;
}

/** OpenAI tool call format */
interface OpenAIToolCall {
  id: string;
  type: 'function';
  function: {
    name: string;
    arguments: string;
  };
}

/** OpenAI tool definition format */
interface OpenAITool {
  type: 'function';
  function: {
    name: string;
    description: string;
    parameters: Record<string, unknown>;
    strict?: boolean;
  };
}

/** OpenAI chat completion response */
interface OpenAIChatCompletion {
  id: string;
  object: string;
  created: number;
  model: string;
  choices: Array<{
    index: number;
    message: {
      role: string;
      content: string | null;
      tool_calls?: OpenAIToolCall[];
    };
    finish_reason: string;
  }>;
  usage: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

// =============================================================================
// OPENAI PROVIDER
// =============================================================================

export class OpenAIProvider implements LLMProvider, LLMProviderWithTools {
  readonly name = 'openai';
  readonly defaultModel = 'gpt-4-turbo-preview';

  private apiKey: string;
  private model: string;
  private baseUrl: string;
  private organization?: string;
  private networkConfig: NetworkConfig;

  constructor(config?: OpenAIConfig) {
    this.apiKey = config?.apiKey ?? requireEnv('OPENAI_API_KEY');
    this.model = config?.model ?? this.defaultModel;
    this.baseUrl = config?.baseUrl ?? 'https://api.openai.com';
    this.organization = config?.organization ?? process.env.OPENAI_ORG_ID;
    this.networkConfig = {
      timeout: 120000,  // 2 minutes
      maxRetries: 3,
      baseRetryDelay: 1000,
    };
  }

  isConfigured(): boolean {
    return hasEnv('OPENAI_API_KEY');
  }

  /**
   * Basic chat without tool support.
   */
  async chat(messages: Message[], options?: ChatOptions): Promise<ChatResponse> {
    const model = options?.model ?? this.model;

    // Convert to OpenAI message format
    const openaiMessages = this.convertMessagesToOpenAIFormat(messages);

    // Build request body
    const body = {
      model,
      messages: openaiMessages,
      max_tokens: options?.maxTokens ?? 4096,
      temperature: options?.temperature ?? 0.7,
      ...(options?.stopSequences && { stop: options.stopSequences }),
    };

    const headers = this.buildHeaders();

    try {
      const { response } = await resilientFetch({
        url: `${this.baseUrl}/v1/chat/completions`,
        init: {
          method: 'POST',
          headers,
          body: JSON.stringify(body),
        },
        providerName: this.name,
        networkConfig: this.networkConfig,
        onRetry: (attempt, delay, error) => {
          console.warn(`[OpenAI] Retry attempt ${attempt} after ${delay}ms: ${error.message}`);
        },
      });

      if (!response.ok) {
        const error = await response.text();
        throw this.handleError(response.status, error);
      }

      const data = await response.json() as OpenAIChatCompletion;
      return this.parseResponse(data);
    } catch (error) {
      if (error instanceof ProviderError) throw error;
      throw new ProviderError(
        `OpenAI request failed: ${(error as Error).message}`,
        this.name,
        'NETWORK_ERROR',
        error as Error
      );
    }
  }

  /**
   * Chat with native tool use support.
   * OpenAI's tool format is our standard format, so minimal conversion needed.
   */
  async chatWithTools(
    messages: (Message | MessageWithContent)[],
    options?: ChatOptionsWithTools
  ): Promise<ChatResponseWithTools> {
    const model = options?.model ?? this.model;

    // Convert messages to OpenAI format
    const openaiMessages = this.convertMessagesWithToolsToOpenAIFormat(messages);

    // Convert tool definitions (already in OpenAI format, just ensure structure)
    const tools = options?.tools?.map(this.convertToolDefinition.bind(this));

    // Build request body
    const body: Record<string, unknown> = {
      model,
      messages: openaiMessages,
      max_tokens: options?.maxTokens ?? 4096,
      temperature: options?.temperature ?? 0.7,
      ...(options?.stopSequences && { stop: options.stopSequences }),
    };

    // Add tools if provided
    if (tools && tools.length > 0) {
      body.tools = tools;

      // Convert tool_choice
      if (options?.tool_choice) {
        body.tool_choice = this.convertToolChoice(options.tool_choice);
      }
    }

    const headers = this.buildHeaders();

    try {
      const { response } = await resilientFetch({
        url: `${this.baseUrl}/v1/chat/completions`,
        init: {
          method: 'POST',
          headers,
          body: JSON.stringify(body),
        },
        providerName: this.name,
        networkConfig: this.networkConfig,
        onRetry: (attempt, delay, error) => {
          console.warn(`[OpenAI] Retry attempt ${attempt} after ${delay}ms: ${error.message}`);
        },
      });

      if (!response.ok) {
        const error = await response.text();
        throw this.handleError(response.status, error);
      }

      const data = await response.json() as OpenAIChatCompletion;
      return this.parseResponseWithTools(data);
    } catch (error) {
      if (error instanceof ProviderError) throw error;
      throw new ProviderError(
        `OpenAI request failed: ${(error as Error).message}`,
        this.name,
        'NETWORK_ERROR',
        error as Error
      );
    }
  }

  // ===========================================================================
  // MESSAGE CONVERSION
  // ===========================================================================

  /**
   * Convert basic messages to OpenAI format.
   */
  private convertMessagesToOpenAIFormat(messages: Message[]): OpenAIMessage[] {
    return messages.map(m => ({
      role: m.role as OpenAIMessage['role'],
      content: m.content,
    }));
  }

  /**
   * Convert messages with potential tool calls to OpenAI format.
   */
  private convertMessagesWithToolsToOpenAIFormat(
    messages: (Message | MessageWithContent)[]
  ): OpenAIMessage[] {
    const result: OpenAIMessage[] = [];

    for (const msg of messages) {
      // Handle tool result messages
      if ('tool_call_id' in msg && msg.role === 'tool') {
        result.push({
          role: 'tool',
          content: typeof msg.content === 'string'
            ? msg.content
            : msg.content?.map(c => c.text).join('') || '',
          tool_call_id: msg.tool_call_id,
          name: msg.name,
        });
        continue;
      }

      // Handle assistant messages with tool calls
      if ('tool_calls' in msg && msg.tool_calls && msg.tool_calls.length > 0) {
        const content = typeof msg.content === 'string'
          ? msg.content
          : msg.content?.map(c => c.text).join('') || null;

        result.push({
          role: 'assistant',
          content,
          tool_calls: msg.tool_calls.map(tc => ({
            id: tc.id,
            type: 'function' as const,
            function: {
              name: tc.function.name,
              arguments: tc.function.arguments,
            },
          })),
        });
        continue;
      }

      // Regular message
      const content = typeof msg.content === 'string'
        ? msg.content
        : msg.content?.map(c => c.text).join('') || '';

      result.push({
        role: msg.role as OpenAIMessage['role'],
        content,
      });
    }

    return result;
  }

  // ===========================================================================
  // TOOL CONVERSION
  // ===========================================================================

  /**
   * Convert tool definition to OpenAI format.
   * Our standard format is already OpenAI-compatible.
   */
  private convertToolDefinition(tool: ToolDefinitionSchema): OpenAITool {
    return {
      type: 'function',
      function: {
        name: tool.function.name,
        description: tool.function.description,
        parameters: tool.function.parameters,
        strict: tool.function.strict,
      },
    };
  }

  /**
   * Convert tool_choice to OpenAI format.
   */
  private convertToolChoice(
    choice: ChatOptionsWithTools['tool_choice']
  ): 'auto' | 'required' | 'none' | { type: 'function'; function: { name: string } } {
    if (choice === 'auto') return 'auto';
    if (choice === 'required') return 'required';
    if (choice === 'none') return 'none';
    if (typeof choice === 'object' && choice.function?.name) {
      return { type: 'function', function: { name: choice.function.name } };
    }
    return 'auto';
  }

  // ===========================================================================
  // RESPONSE PARSING
  // ===========================================================================

  /**
   * Parse basic response without tools.
   */
  private parseResponse(data: OpenAIChatCompletion): ChatResponse {
    const choice = data.choices[0];

    return {
      content: choice.message.content || '',
      stopReason: this.mapStopReason(choice.finish_reason),
      usage: {
        inputTokens: data.usage.prompt_tokens,
        outputTokens: data.usage.completion_tokens,
      },
    };
  }

  /**
   * Parse response that may include tool calls.
   */
  private parseResponseWithTools(data: OpenAIChatCompletion): ChatResponseWithTools {
    const choice = data.choices[0];

    // Convert OpenAI tool_calls to our format
    const toolCalls: ToolCallResponse[] | undefined = choice.message.tool_calls?.map(tc => ({
      id: tc.id,
      type: 'function' as const,
      function: {
        name: tc.function.name,
        arguments: tc.function.arguments,
      },
    }));

    return {
      content: choice.message.content || '',
      stopReason: this.mapStopReason(choice.finish_reason),
      usage: {
        inputTokens: data.usage.prompt_tokens,
        outputTokens: data.usage.completion_tokens,
      },
      toolCalls: toolCalls && toolCalls.length > 0 ? toolCalls : undefined,
    };
  }

  // ===========================================================================
  // UTILITIES
  // ===========================================================================

  /**
   * Build request headers.
   */
  private buildHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${this.apiKey}`,
    };

    if (this.organization) {
      headers['OpenAI-Organization'] = this.organization;
    }

    return headers;
  }

  /**
   * Handle API errors.
   */
  private handleError(status: number, body: string): ProviderError {
    let code: ProviderError['code'] = 'UNKNOWN';

    if (status === 401) code = 'AUTHENTICATION_FAILED';
    else if (status === 429) code = 'RATE_LIMITED';
    else if (status === 400) {
      if (body.includes('context_length') || body.includes('maximum context length')) {
        code = 'CONTEXT_LENGTH_EXCEEDED';
      } else {
        code = 'INVALID_REQUEST';
      }
    }
    else if (status === 404) code = 'INVALID_REQUEST';
    else if (status >= 500) code = 'SERVER_ERROR';

    return new ProviderError(
      `OpenAI API error (${status}): ${body}`,
      this.name,
      code
    );
  }

  /**
   * Map OpenAI stop reasons to our format.
   */
  private mapStopReason(reason: string): ChatResponse['stopReason'] {
    switch (reason) {
      case 'stop': return 'end_turn';
      case 'length': return 'max_tokens';
      case 'tool_calls': return 'end_turn'; // Tool calls treated as end_turn
      case 'content_filter': return 'end_turn';
      default: return 'end_turn';
    }
  }
}

// =============================================================================
// REGISTRATION
// =============================================================================

registerProvider('openai', {
  priority: 2,
  detect: () => hasEnv('OPENAI_API_KEY'),
  create: async () => new OpenAIProvider(),
});

File: attocode/attocode_py/src/attocode/providers/openai.py
"""OpenAI API provider using httpx."""

from __future__ import annotations

import json
import os
from collections.abc import AsyncIterator
from typing import Any

import httpx

from attocode.errors import ProviderError
from attocode.types.messages import (
    ChatOptions,
    ChatResponse,
    Message,
    MessageWithStructuredContent,
    Role,
    StopReason,
    StreamChunk,
    TokenUsage,
    ToolCall,
    ToolDefinition,
)

DEFAULT_API_URL = "https://api.openai.com/v1/chat/completions"
DEFAULT_MODEL = "gpt-4o"


def _describe_request_error(e: httpx.RequestError) -> str:
    """Build a descriptive error message for httpx request errors.

    httpx.ReadTimeout and similar errors often have empty str(e),
    so we fall back to the exception type name and include the
    chained cause when available.
    """
    msg = str(e)
    if not msg:
        msg = type(e).__name__
    if e.__cause__ and str(e.__cause__):
        msg = f"{msg} (caused by {type(e.__cause__).__name__}: {e.__cause__})"
    return msg


class OpenAIProvider:
    """OpenAI API provider using httpx."""

    def __init__(
        self,
        api_key: str | None = None,
        model: str = DEFAULT_MODEL,
        api_url: str = DEFAULT_API_URL,
        timeout: float = 600.0,
    ) -> None:
        self._api_key = api_key or os.environ.get("OPENAI_API_KEY", "")
        if not self._api_key:
            raise ProviderError("OPENAI_API_KEY not set", provider="openai", retryable=False)
        self._model = model
        self._api_url = api_url
        self._timeout = timeout
        self._client = self._create_client()

    def _create_client(self) -> httpx.AsyncClient:
        """Create a fresh httpx client."""
        return httpx.AsyncClient(
            timeout=httpx.Timeout(self._timeout),
            headers={"Authorization": f"Bearer {self._api_key}", "Content-Type": "application/json"},
        )

    def _ensure_client(self) -> httpx.AsyncClient:
        """Return the client, recreating it if closed."""
        if self._client.is_closed:
            self._client = self._create_client()
        return self._client

    @property
    def name(self) -> str:
        return "openai"

    async def chat(
        self,
        messages: list[Message | MessageWithStructuredContent],
        options: ChatOptions | None = None,
    ) -> ChatResponse:
        client = self._ensure_client()
        model = (options and options.model) or self._model
        body: dict[str, Any] = {"model": model, "messages": self._format_messages(messages)}
        if options and options.max_tokens:
            body["max_tokens"] = options.max_tokens
        if options and options.temperature is not None:
            body["temperature"] = options.temperature
        if options and options.tools:
            body["tools"] = [self._format_tool(t) for t in options.tools]

        try:
            response = await client.post(self._api_url, json=body)
            response.raise_for_status()
            return self._parse_response(response.json(), model)
        except httpx.HTTPStatusError as e:
            status = e.response.status_code
            raise ProviderError(
                f"OpenAI API error {status}: {e.response.text[:500]}",
                provider="openai", status_code=status, retryable=status in (429, 500, 502, 503),
            ) from e
        except httpx.RequestError as e:
            raise ProviderError(
                f"OpenAI request error: {_describe_request_error(e)}",
                provider="openai", retryable=True,
            ) from e

    async def chat_stream(
        self,
        messages: list[Message | MessageWithStructuredContent],
        options: ChatOptions | None = None,
    ) -> AsyncIterator[StreamChunk]:
        """Stream a chat response from the OpenAI API."""
        from attocode.integrations.streaming.handler import adapt_openrouter_stream

        client = self._ensure_client()
        model = (options and options.model) or self._model
        body: dict[str, Any] = {
            "model": model,
            "messages": self._format_messages(messages),
            "stream": True,
            "stream_options": {"include_usage": True},
        }
        if options and options.max_tokens:
            body["max_tokens"] = options.max_tokens
        if options and options.temperature is not None:
            body["temperature"] = options.temperature
        if options and options.tools:
            body["tools"] = [self._format_tool(t) for t in options.tools]

        try:
            async with client.stream("POST", self._api_url, json=body) as response:
                response.raise_for_status()
                async for chunk in adapt_openrouter_stream(response.aiter_lines()):
                    yield chunk
        except httpx.HTTPStatusError as e:
            status = e.response.status_code
            raise ProviderError(
                f"OpenAI API error {status}: {e.response.text[:500]}",
                provider="openai", status_code=status, retryable=status in (429, 500, 502, 503),
            ) from e
        except httpx.RequestError as e:
            raise ProviderError(
                f"OpenAI request error: {_describe_request_error(e)}",
                provider="openai", retryable=True,
            ) from e
        except httpx.StreamError as e:
            raise ProviderError(
                f"OpenAI stream error: {type(e).__name__}: {e}",
                provider="openai", retryable=True,
            ) from e

    def _format_messages(self, messages: list[Message | MessageWithStructuredContent]) -> list[dict[str, Any]]:
        result: list[dict[str, Any]] = []
        for msg in messages:
            content = msg.content if isinstance(msg.content, str) else str(msg.content)
            if msg.role == Role.TOOL:
                result.append({"role": "tool", "content": content, "tool_call_id": msg.tool_call_id or ""})
            elif msg.role == Role.ASSISTANT and msg.tool_calls:
                tc_list = [
                    {"id": tc.id, "type": "function", "function": {"name": tc.name, "arguments": json.dumps(tc.arguments)}}
                    for tc in msg.tool_calls
                ]
                result.append({"role": "assistant", "content": content or None, "tool_calls": tc_list})
            else:
                result.append({"role": str(msg.role), "content": content})
        return result

    def _format_tool(self, tool: ToolDefinition) -> dict[str, Any]:
        return {"type": "function", "function": {"name": tool.name, "description": tool.description, "parameters": tool.parameters}}

    def _parse_response(self, data: dict[str, Any], model: str) -> ChatResponse:
        choices = data.get("choices", [])
        if not choices:
            return ChatResponse(content="", model=model, stop_reason=StopReason.END_TURN)
        choice = choices[0]
        message = choice.get("message", {})
        content = message.get("content", "") or ""

        tool_calls: list[ToolCall] | None = None
        if raw_tc := message.get("tool_calls"):
            tool_calls = []
            for tc in raw_tc:
                func = tc.get("function", {})
                try:
                    args = json.loads(func.get("arguments", "{}"))
                except json.JSONDecodeError:
                    args = {}
                tool_calls.append(ToolCall(id=tc["id"], name=func["name"], arguments=args))

        usage_data = data.get("usage", {})
        usage = TokenUsage(
            input_tokens=usage_data.get("prompt_tokens", 0),
            output_tokens=usage_data.get("completion_tokens", 0),
            total_tokens=usage_data.get("total_tokens", 0),
        )

        finish = choice.get("finish_reason", "stop")
        stop_reason = StopReason.TOOL_USE if finish == "tool_calls" else (StopReason.MAX_TOKENS if finish == "length" else StopReason.END_TURN)

        return ChatResponse(content=content, tool_calls=tool_calls, usage=usage, model=model, stop_reason=stop_reason)

    async def close(self) -> None:
        await self._client.aclose()
